internal/engine/planner/lambda.go

```diff
@@ -4,54 +4,156 @@
+		iterRange: e.AccuInit,
+		iterVar:   e2.IterVar,
+		expr:      (*wrapper)(e2.LoopStep).getArg(1).getListElement(0).e(),
+	}
+	if obj.expr == nil {
+		return nil, fmt.Errorf("%w, got %s", ErrExpectedSortBy, e.String())
+	}
+	return obj, nil
 }
 
 func buildLambdaAST(e *exprpb.Expr_Comprehension) (*lambdaAST, error) {
+	var function string
+	var loopStep *wrapper
+	switch ls := e.LoopStep.ExprKind.(type) {
+	case *exprpb.Expr_CallExpr:
+		function = ls.CallExpr.Function
+		loopStep = (*wrapper)(e.LoopStep)
+	case *exprpb.Expr_IdentExpr:
+		return mkSortByAST(e)
+	default:
+		return nil, fmt.Errorf("expected loop-step expression type CallExpr, got: %T", ls)
+	}
 	obj := &lambdaAST{
 		iterVar:   e.IterVar,
+		iterVar2:  e.IterVar2,
 		iterRange: e.IterRange,
 	}
-	var step *exprpb.Expr_CallExpr
-	var ok bool
-	if step, ok = e.LoopStep.ExprKind.(*exprpb.Expr_CallExpr); !ok {
-		return nil, fmt.Errorf("expected loop-step expression type CallExpr, got: %T", e.LoopStep.ExprKind)
-	}
-	switch step.CallExpr.Function {
+	switch function {
 	case operators.LogicalAnd:
 		obj.operator = All
-		obj.lambdaExpr = step.CallExpr.Args[1]
+		obj.expr = loopStep.getArg(1).e()
 	case operators.LogicalOr:
 		obj.operator = Exists
-		obj.lambdaExpr = step.CallExpr.Args[1]
+		obj.expr = loopStep.getArg(1).e()
 	case operators.Add:
 		obj.operator = Map
-		if elements := step.CallExpr.Args[1].GetListExpr().GetElements(); len(elements) > 0 {
-			obj.lambdaExpr = elements[0]
+		if obj.iterVar2 != "" {
+			obj.operator = TransformList
 		}
+		obj.expr = loopStep.getArg(1).getListElement(0).e()
 	case operators.Conditional:
 		switch e.AccuInit.ExprKind.(type) {
 		case *exprpb.Expr_ListExpr:
-			obj.operator = Filter
+			if e2 := loopStep.getArg(1).getArg(1).getListElement(0).e(); e2.GetCallExpr() != nil {
+				obj.expr2 = e2
+				obj.operator = TransformList
+			} else {
+				obj.operator = Filter
+			}
 		case *exprpb.Expr_ConstExpr:
 			obj.operator = ExistsOne
+		case *exprpb.Expr_StructExpr:
+			var err error
+			obj.operator, obj.expr2, err = getTransformMapExpression(loopStep.getArg(1))
+			if err != nil {
+				return nil, err
+			}
 		default:
 			return nil, fmt.Errorf("expected loop-accu-init expression type ConstExpr or ListExpr, got: %T", e.AccuInit.ExprKind)
 		}
-		obj.lambdaExpr = step.CallExpr.Args[0]
+		obj.expr = loopStep.getArg(0).e()
+	case "cel.@mapInsert":
+		var err error
+		obj.operator, obj.expr, err = getTransformMapExpression(loopStep)
+		if err != nil {
+			return nil, err
+		}
 	default:
-		return nil, fmt.Errorf("unexpected loop-step function: %q", step.CallExpr.Function)
+		return nil, fmt.Errorf("unexpected loop-step function: %q", function)
 	}
 
 	return obj, nil
 }
+
+type wrapper exprpb.Expr
+
+func (w *wrapper) e() *exprpb.Expr {
+	return (*exprpb.Expr)(w)
+}
+
+func (w *wrapper) getArg(i int) *wrapper {
+	if x := w.e().GetCallExpr(); x != nil && i < len(x.Args) {
+		return (*wrapper)(x.Args[i])
+	}
+	return nil
+}
```

Comment:
One way to handle the nil issues pointed out above would be to do a nil check at the beginning of each method call provided by the `wrapper` object.
```suggestion
func (w *wrapper) getArg(i int) *wrapper {
    if w == nil {
        return nil
    }
	if x := w.e().GetCallExpr(); x != nil && i < len(x.Args) {
		return (*wrapper)(x.Args[i])
	}
	return nil
}
```

Reply:
I had these checks when a wrapper was a struct, but currently, it is a simple type alias of the protobuf `Expr` type, so the `nil` checks aren't needed. The 'wrapper' name may not fit here.
`w.e()` casts `*wrapper` to `*Expr`, then calls protobuf functions, which correctly handle `nil` receivers.

The `TestWrapperDoNotPanic` test, hopefully, proves it.